torchtitan/optimizer.py

```diff
@@ -38,6 +38,10 @@ def __init__(
             else:
                 raise NotImplementedError(f"Optimizer {name} not added.")
             self.optimizers.append(optimizer)
+        expected_optim_num = len(self.model_parts)
+        assert expected_optim_num == len(
+            self.optimizers
+        ), "Must pass one optimizer per model part"
```

Comment:
how about let's isolate this assertion into a private function `_validate_length(expected_length)` in the base class, so that we can reuse it in any optimizer class